sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldHaveCorrectSignature.cs

```diff
@@ -0,0 +1,124 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class TestMethodShouldHaveCorrectSignature : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S3433";
+        private const string MessageFormat = "Make this test method {0}.";
+        private const string MakePublicMessage = "'public'";
+        private const string MakeNonAsyncMessage = "non-'async'";
+        private const string MakeNotGenericMessage = "non-generic";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly ISet<KnownType> TrackedTestAttributes = ImmutableHashSet.Create(
+            KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_TestMethodAttribute,
+            KnownType.NUnit_Framework_TestAttribute,
+            KnownType.Xunit_FactAttribute);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    if (!c.IsTest())
+                    {
+                        return;
+                    }
+
+                    var classDeclaration = (ClassDeclarationSyntax)c.Node;
+                    var classSymbol = c.SemanticModel.GetDeclaredSymbol(classDeclaration);
+                    if (classSymbol == null)
+                    {
+                        return;
+                    }
+
+                    var allFaultyMethods = classSymbol.GetMembers()
+                        .OfType<IMethodSymbol>()
+                        .Where(method => method.GetAttributes()
+                            .Any(attribute => attribute.AttributeClass.IsAny(TrackedTestAttributes)))
+                        .Select(method => new
+                        {
+                            Location = method.Locations.First(),
```

Comment:
Could method.Locations possibly be null or empty? Perhaps worth going defensive with method.Locations?.FirstOrDefault() ?